### PR TITLE
Spellchecking added

### PR DESCRIPTION
**Pull request by thecyberd3m0n**
Tested on Linux and Windows.

----

Fixes https://github.com/vector-im/riot-web/issues/2661 ~ Travis

### CODE REVIEW

File: electron_app/src/preload.js [hunk 1]
```diff
@@ -15,8 +15,9 @@ limitations under the License.
 */
 
 const { ipcRenderer, webFrame } = require('electron');
-
 // expose ipcRenderer to the renderer process
+// const SpellCheck = require('./spell_check');
```
Comment (MTRNord) [inline electron_app/src/preload.js]: This comment should get removed

File: electron_app/src/spell-check.js [hunk 2]
```diff
@@ -0,0 +1,309 @@
...
+			return null;
+		}
+	})();
+
+	if (selectedDictionaries) {
+		enable(...selectedDictionaries);
+		return;
+	}
+
+	const userLanguage = localStorage.getItem('userLanguage');
```
Comment (MTRNord) [inline electron_app/src/spell-check.js]: About this whole block: Can't you use the language the User selected inside the app using the already existing code?

File: .vscode/launch.json [hunk 3]
```diff
@@ -0,0 +1,35 @@
+{
```
Comment (anoadragon453) [inline .vscode/launch.json]: This file is quite specific to your setup/filesystem. Not sure it needs to be checked in unless you can make it generic.

File: electron_app/src/electron-main.js [hunk 4]
```diff
@@ -322,14 +322,13 @@ app.on('ready', () => {
         webPreferences: {
             preload: preloadScript,
             nodeIntegration: false,
-            sandbox: true,
```
Comment (anoadragon453) [inline electron_app/src/electron-main.js]: Removing this sounds slightly scary https://electronjs.org/docs/api/sandbox-option

Is it possible to make this all work with this still enforcing?

**Comment by thecyberd3m0n** [comment]
I'll sit resolving issue in ~5hours

File: .vscode/launch.json (hunk 3 above)
Comment (thecyberd3m0n) [reply .vscode/launch.json]: I'll add .vscode to .gitignore ok?

File: electron_app/src/spell-check.js (hunk 2 above)
Comment (thecyberd3m0n) [reply electron_app/src/spell-check.js]: I'll try ;)

File: electron_app/src/electron-main.js (hunk 4 above)
Comment (thecyberd3m0n) [reply electron_app/src/electron-main.js]: image
After enabling Sandbox errors like this happened. Could you help me figure this out?

Comment (anoadragon453) [reply electron_app/src/electron-main.js]: (With my limited knowledge of Electron - I don't work on Riot) perhaps the sandbox prevents some parts of the code from communicating directly to modules, and instead you need to use IPC communication to wrap calls to the lib.

@dbkr would know more though.

Comment (thecyberd3m0n) [reply electron_app/src/electron-main.js]: ok, I'll try to figure it out today

**Comment by turt2live** [comment]
Please also use spaces instead of tabs, per the [code style](https://github.com/matrix-org/matrix-js-sdk/blob/master/CONTRIBUTING.rst#code-style). We'll also need signoff before we can merge this.

File: electron_app/src/electron-main.js (hunk 4 above)
Comment (thecyberd3m0n) [reply electron_app/src/electron-main.js]: ok I think it's fixed

**Comment by thecyberd3m0n** [comment]
code style fixed due to eslint

**Comment by thecyberd3m0n** [comment]
please wait because installing dictionaries not works yet.

**Comment by thecyberd3m0n** [comment]
ok, if someone could test on Mac, while I'll configure my Mac VM for future contributions?:) I can't build for Mac without Mac.

File: electron_app/src/electron-main.js [hunk 5]
```diff
@@ -322,14 +323,14 @@ app.on('ready', () => {
...
-            enableRemoteModule: false,
             // We don't use this: it's useful for the preload script to
             // share a context with the main page so we can give select
             // objects to the main page. The sandbox option isolates the
             // main page from the background script.
+            sandbox: true,
```
Comment (anoadragon453) [inline electron_app/src/electron-main.js]: Could you move this back to where it was to make the diff cleaner?

Comment (thecyberd3m0n) [reply electron_app/src/electron-main.js]: ok

File: electron_app/src/electron-main.js [hunk 6]
```diff
@@ -321,15 +322,15 @@ app.on('ready', () => {
         height: mainWindowState.height,
         webPreferences: {
             preload: preloadScript,
-            nodeIntegration: false,
             sandbox: true,
-            enableRemoteModule: false,
+            nodeIntegration: false,
```
Comment (anoadragon453) [inline electron_app/src/electron-main.js]: This one too? :)

Comment (thecyberd3m0n) [reply electron_app/src/electron-main.js]: enableRemoteModule: false leads to errors in my case. Can I leave it?

Comment (thecyberd3m0n) [reply electron_app/src/electron-main.js]: image

Comment (thecyberd3m0n) [reply electron_app/src/electron-main.js]: @anoadragon453 ?

Comment (anoadragon453) [reply electron_app/src/electron-main.js]: Sorry, I meant just moving it back to where it was originally in the list, not changing the value.

Comment (thecyberd3m0n) [reply electron_app/src/electron-main.js]: fixed

File: electron_app/src/electron-main.js [hunk 7]
```diff
@@ -323,13 +324,14 @@ app.on('ready', () => {
...
             sandbox: true,
-            enableRemoteModule: false,
+            enableRemoteModule: true,
             // We don't use this: it's useful for the preload script to
             // share a context with the main page so we can give select
             // objects to the main page. The sandbox option isolates the
             // main page from the background script.
             contextIsolation: false,
             webgl: false,
+            nativeWindowOpen: true,
```
Comment (t3chguy) [inline electron_app/src/electron-main.js]: note: this option is experimental
See https://electronjs.org/docs/api/browser-window

File: electron_app/yarn-error.log [hunk 8]
```diff
@@ -0,0 +1,905 @@
+Arguments:
```
Comment (t3chguy) [inline electron_app/yarn-error.log]: This file should not be staged

Comment (thecyberd3m0n) [reply electron_app/yarn-error.log]: good point, I'll .gitignore it

File: electron_app/src/electron-main.js (hunk 7 above)
Comment (thecyberd3m0n) [reply electron_app/src/electron-main.js]: I'll test feature without it

Comment (thecyberd3m0n) [reply electron_app/src/electron-main.js]: ok it works without it. Sorry for commiting it, I was inspired by Rocket Chat implementation, so I copied their settings ;)

File: electron_app/yarn-error.log (hunk 8 above)
Comment (thecyberd3m0n) [reply electron_app/yarn-error.log]: fixed

File: electron_app/src/spell-check.js [hunk 9]
```diff
@@ -0,0 +1,217 @@
+/*
+Copyright 2018 New Vector Ltd
```
Comment (turt2live) [inline electron_app/src/spell-check.js]: We can't take copyright for this. It looks almost identical to Rocket.Chat's implementation (https://github.com/RocketChat/Rocket.Chat.Electron/blob/7ad28f3f24f7b32b22ea75ed4e35ed2b173b933c/src/preload/spellchecking.js) which is under MIT. We need to give them credit, and use their license for this file, in order to include it.

File: electron_app/yarn-error.log (hunk 8 above)
Comment (dbkr) [reply electron_app/yarn-error.log]: It shouldn't really be ignored here either: it's nothing to do with the functionality this PR is adding

**Comment by dbkr** [comment]
I haven't exhaustively gone through this yet, but a few general points:

 1. See @turt2live 's comment about copyright

 1. Your indenting is all over the place: 4 space indents please (https://github.com/matrix-org/matrix-react-sdk/blob/develop/code_style.md)

 1. There isn't a single comment in this PR and there's a lot of code that needs explanation. For example...

 1. ...it looks like you're installing new dictionary files to the app directory? Won't these be wiped away when the app updates and/or the user won't have permission to write there?

 1. I really, really, *really* dislike having dictionary files in the riot codebase. Where did they come from? How would we update them? Is whatever `getAvailableDictionaries()` has by default not enough?

 1. You're enabling the remote module which is *probably* okay security-wise, but it looks like it's mostly used to show context menus from the renderer process. Why not just do it from the main process?

 1. You're adding a lot of dependencies (eg. jetpack, mem, os (is this one even used?)) This might be ok but it's not really clear that they're necessarily justifying their presence.

 1. and mostly: the spellchecker module is a native module. We're still looking at the implications of adding native modules to Riot but basically we don't yet have enough infrastructure to build them ourselves and for a security-sensitive encrypted messenger, pulling in more precompiled binaries from more not-necessarily-trusted sources is somewhat nonideal.

Thanks for looking at this, it is a somewhat complicated one though I'm afraid.

File: electron_app/src/spell-check.js (hunk 9 above)
Comment (thecyberd3m0n) [reply electron_app/src/spell-check.js]: we could change something inside. For example their implementation allow Multilanguage, our not ;)
Please suggest something

**Comment by thecyberd3m0n** [comment]
I have inspired by working Rocket.Chat implementation, so similarities can be found.

1. I'll change the implementation, by removing dependencies and replacing it by Node.js vanilla implementations (also 7 - that will be fixed)
2. I used eslint, I'll recheck it.
3. I'll add some in new code
4, 5
6. You know I tried several hours but without success. It was a workaround. I could retry because of new conditions.

8. Maybe I'll replace it by https://github.com/atom/node-spellchecker

**Comment by thecyberd3m0n** [comment]
@dbkr can I add 4 spaces rule to eslintrc?

File: electron_app/src/spell-check.js (hunk 9 above)
Comment (turt2live) [reply electron_app/src/spell-check.js]: I'm not a lawyer, but I believe we need to have the file under MIT and their copyright. Then when changes are made the applicable party's copyright is added underneath the Rocket.Chat copyright.

**Comment by thecyberd3m0n** [comment]
@dbkr 
I resolved most of issues. In details:
1) I think it's not the case according to new version
2) I've added rule to eslint and fixed automatically using IDE
3) there was one place where it was reasonable
4) 5) simple-spellchecker also comes with dictionaries out of box. Please give some ideas how we can adddress this issue. How we should install dictionaries, from where app should get them? Please help I'm running out of ideas
6) not the case anymore, that part was left untouched
7) now just simple-spellchecker should be new
8) simple-spellchecker is not a native module ;)

File: electron_app/src/spell-check.js (hunk 9 above)
Comment (thecyberd3m0n) [reply electron_app/src/spell-check.js]: Please take a look at new version

File: electron_app/src/preload.js [hunk 10]
```diff
@@ -15,10 +15,12 @@ limitations under the License.
 */
 
 const { ipcRenderer, webFrame } = require('electron');
-
-// expose ipcRenderer to the renderer process
```
Comment (MTRNord) [inline electron_app/src/preload.js]: Why did you remove this comment?

Comment (thecyberd3m0n) [reply electron_app/src/preload.js]: sorry;)

**Comment by thecyberd3m0n** [comment]
did some diff cleanup. please check again. Perhaps I trust automated tools for code-style too much

**Comment by dbkr** [comment]
> 1. I think it's not the case according to new version

The copyright? Rocketchat is MIT licensed, so it's fine to use their code, but we must attribute it appropriately. Code you write is copyright yourself (ie. not New Vector).

> 2. I've added rule to eslint and fixed automatically using IDE

See inline comment, but you've also introduced a lot of unrelated formatting changes.

> 4. 1. simple-spellchecker also comes with dictionaries out of box. Please give some ideas how we can adddress this issue. How we should install dictionaries, from where app should get them? Please help I'm running out of ideas

Honestly I'm not sure - I'd have to go and look at what the various parts of the spellchecking infrastructure come with. Unfortunately this is one of the reasons this PR is tricky, ie. it's not just a question of writing the code.

Thanks again for the work on this, although I think there's still a lot to be done: I'm not really sure I understand how this code works and there still aren't really many comments to help me. As an example, what do we gain/lose from simple-spellcheck over what you were using before? Presumably it's no longer using the OS spellchecking so words the user added to their system's dictionary won't be included?

You've certainly picked a very complex issue to fix: we'd normally expect people to start with a simpler one for their first PR - you could maybe start with some issues with the 'easy' tag?

**Comment by thecyberd3m0n** [comment]
maybe. But if we've got time we can investigate issues here, and build to the end. If not you can close PR, I'll focus on other issues :)

2) I'll try to minize the diff, sorry for that ;)

simple-spellcheck is found by me replacement over native module, that comes with node-spellchecker and electron-spellchecker (https://www.npmjs.com/package/simple-spellchecker). Unfortunately - it bundles dictionaries inside. It also allows to install dictionaries.
Simple-spellcheck should not require any major infrastructure upgrade, as it uses Javascript code https://github.com/jfmdev/simple-spellchecker.

**Comment by jryans** [comment]
The PR author mentioned in the issue that the PR can be closed. Thanks for investigating this issue, and I'm sure we'll find a fix down the road.
